### PR TITLE
2234/ability to clear whole cart

### LINKED ISSUE #2234: Checkout process improvements
During my work on PR #2206 I found out few things that may improve checkout process.

#### Going forward/back between steps

Although we can go back to shipping address step from choosing shipping method, we can't go back from summary step, where billing address is set. I think that passing between steps should be allowed.

#### Clearing cart

We can remove every single item from cart, but there is no button like `Clear cart`. It would be useful, when customer wants to cancel all cart lines at once.

### PR DESCRIPTION
**Pull request by tomcio1205**
close #2234 

### Pull Request Checklist

1. [x] Privileged views and APIs are guarded by proper permission checks.
1. [x] All visible strings are translated with proper context.
1. [x] All data-formatting is locale-aware (dates, numbers, and so on).
1. [x] Database queries are optimized and the number of queries is constant.
1. [x] The changes are tested.
1. [x] The code is documented (docstrings, project documentation).

### CODE REVIEW

File: saleor/checkout/views/__init__.py [hunk 1]
```diff
@@ -195,6 +195,25 @@ def update_cart_line(request, cart, variant_id):
...
 
+@get_or_empty_db_cart()
+def clear_cart(request, cart):
+    """Clear cart"""
+    from ..utils import update_cart_quantity
+
+    if not request.is_ajax():
+        return redirect('cart:index')
+
+    lines = cart.lines.all()
```
Comment (NyanKiyoshi) [inline saleor/checkout/views/__init__.py]: You could do this as well:
```python
cart.lines.all().delete()
```

File: saleor/checkout/views/__init__.py [hunk 2]
```diff
@@ -195,6 +195,25 @@ def update_cart_line(request, cart, variant_id):
...
+    for line in lines:
+        line.delete()
+
+    update_cart_quantity(cart)
+
+    response = {'numItems': 0}
+
+    return JsonResponse(response, status=200)
```
Comment (Pacu2) [inline saleor/checkout/views/__init__.py]: As far as I remember default status is `200`

Also, this function is stretched up a little, we could tighten it up :)

File: saleor/static/js/components/cart.js [hunk 3]
```diff
@@ -57,6 +57,19 @@ export default $(document).ready((e) => {
...
+  $('.cart__clear').click((e) => {
+    $.ajax({
+      url: $('.cart__clear').attr('action'),
+      method: 'POST',
+      data: {},
+      success: (response) => {
+        $('.badge').html(response.numItems)
+        $.cookie('alert', 'true', {path: '/cart'});
+        location.reload();
+        deliveryAjax();
```
Comment (NyanKiyoshi) [inline saleor/static/js/components/cart.js]: Is it really necessary? In addition, it is called after instructing the browser to reload the page.

File: saleor/checkout/views/__init__.py [hunk 4]
```diff
@@ -195,6 +195,25 @@ def update_cart_line(request, cart, variant_id):
...
+@get_or_empty_db_cart()
+def clear_cart(request, cart):
+    """Clear cart"""
+    from ..utils import update_cart_quantity
+
+    if not request.is_ajax():
+        return redirect('cart:index')
+
+    lines = cart.lines.all()
+    for line in lines:
```
Comment (Pacu2) [inline saleor/checkout/views/__init__.py]: We can iterate over the `cart` object here, as it's `__iter__` method is overridden especially for such purpose 
https://github.com/mirumee/saleor/blob/master/saleor/checkout/models.py#L70-L71

File: saleor/static/scss/layouts/_cart.scss [hunk 5]
```diff
@@ -100,4 +100,7 @@
       width: 100%;
     }
   }
+  &__clear{
+    margin-right: 10px;
```
Comment (NyanKiyoshi) [inline saleor/static/scss/layouts/_cart.scss]: You could use a bootstrap' spacing utility classes instead, it's also a really bad idea to use pixels. (https://getbootstrap.com/docs/4.1/utilities/spacing/)

File: templates/checkout/index.html [hunk 6]
```diff
@@ -80,6 +80,9 @@
           <a href="{% url "checkout:login" %}" class="btn primary float-right cart__submit">
             {% trans "Checkout" context "Cart primary action" %}
           </a>
+          <button class="btn secondary float-right cart__clear" action="{% url 'cart:clear-cart' %}">
+            {% trans "Clear Cart" context "Clear Cart" %}
```
Comment (NyanKiyoshi) [inline templates/checkout/index.html]: Could you add a more meaningful context? Could be "Cart summary page: secondary action, empty the cart".

File: templates/checkout/summary.html [hunk 7]
```diff
@@ -7,6 +7,7 @@
     {% csrf_token %}
     <h2>{% trans "Billing address" context "Checkout summary title" %}</h2>
     {% include "checkout/snippets/addresses_form.html" with addresses_form=addresses_form addresses=additional_addresses address_form=address_form only %}
+    <p><a href="{% url 'checkout:shipping-method' %}">{% trans "Change shipping method" %}</a></p>
```
Comment (NyanKiyoshi) [inline templates/checkout/summary.html]: Missing context here.

File: templates/checkout/index.html [hunk 8]
```diff
@@ -80,6 +80,9 @@
           <a href="{% url "checkout:login" %}" class="btn primary float-right cart__submit">
             {% trans "Checkout" context "Cart primary action" %}
           </a>
+          <button class="btn secondary float-right cart__clear" action="{% url 'cart:clear-cart' %}">
```
Comment (NyanKiyoshi) [inline templates/checkout/index.html]: You should use `data-action` instead (thus, do `.data('action')` instead).

File: saleor/checkout/views/__init__.py [hunk 9]
```diff
@@ -195,6 +195,25 @@ def update_cart_line(request, cart, variant_id):
     return JsonResponse(response, status=status)
 
 
+@get_or_empty_db_cart()
+def clear_cart(request, cart):
+    """Clear cart"""
+    from ..utils import update_cart_quantity
```
Comment (NyanKiyoshi) [inline saleor/checkout/views/__init__.py]: It's safe to import it at the top of the file (accidental local import?)

Comment (tomcio1205) [reply saleor/checkout/views/__init__.py]: yeah, thats true. I always place import statement on top of file, but i saw this kind of import in another file so i think that this is proper way in this project. But good to know that I was wrong:)

File: tests/test_cart.py [hunk 10]
```diff
@@ -654,3 +654,20 @@ def test_cart_taxes(request_cart_with_item, shipping_method, vatlayer):
...
+
+def test_clear_cart_must_be_ajax(rf, customer_user):
+    request = rf.post(reverse('home'))
+    request.user = customer_user
+    request.discounts = None
+    response = clear_cart(request)
+    assert response.status_code == 302
+
+
+def test_clear_cart_check_if_cart_is_empty(request_cart_with_item, rf, client):
```
Comment (Pacu2) [inline tests/test_cart.py]: I think we should just call it `test_clear_cart` as it tests the expected behavior.

Also rf seems unused in this function

File: templates/checkout/index.html [hunk 11]
```diff
@@ -80,6 +80,9 @@
           <a href="{% url "checkout:login" %}" class="btn primary float-right cart__submit">
             {% trans "Checkout" context "Cart primary action" %}
           </a>
+          <button class="btn secondary float-right mr-3 cart__clear" data-action="{% url 'cart:clear-cart' %}">
+            {% trans "Clear Cart" context "cart page: secondary action, empty the cart" %}
```
Comment (maarcingebala) [inline templates/checkout/index.html]: Correct spelling is "Clear cart".

**Comment by maarcingebala** [comment]
@tomcio1205 Could you provide a screenshot of how the cart page looks after your changes? Also, have you checked if it looks properly on mobile screen resolutions?
Second thing, could you edit your commit messages to follow [these rules](https://github.com/erlang/otp/wiki/writing-good-commit-messages)? Now they're a written in a bit inconsistent manner (e.g. capitalization).

**Comment by tomcio1205** [comment]
@maarcingebala I added few css rules because it looked not very well so, please check if now code and view looks fine.
cart1
cart2
cart3

**Comment by maarcingebala** [comment]
@tomcio1205 It looks good on desktop, but on mobile there should be some spacing between those buttons. In `variables.scss` we have some global spacers that we use across the project like `$global-margin` or `$global-padding`.

File: templates/checkout/summary.html [hunk 12]
```diff
@@ -7,6 +7,7 @@
     {% csrf_token %}
     <h2>{% trans "Billing address" context "Checkout summary title" %}</h2>
     {% include "checkout/snippets/addresses_form.html" with addresses_form=addresses_form addresses=additional_addresses address_form=address_form only %}
+    <p><a href="{% url 'checkout:shipping-method' %}">{% trans "Change shipping method" context "Checkout shipping method" %}</a></p>
```
Comment (maarcingebala) [inline templates/checkout/summary.html]: Is this change related to this PR?

Comment (tomcio1205) [reply templates/checkout/summary.html]: @maarcingebala Yes It is. Issue #2234 also including ability for returning from summary page to shiping-method page, so I added this simple link. It looks as you can see on screenshot. Also I corrected button spacing on mobile view.
cart1
cart2

**Comment by maarcingebala** [comment]
Looks good now! Congrats on your first PR merged to Saleor and thanks a lot for your effort!
